homeassistant/components/tessie/coordinator.py

```diff
@@ -0,0 +1,84 @@
+"""Tessie Data Coordinator."""
+from datetime import timedelta
+from http import HTTPStatus
+import logging
+from typing import Any
+
+from aiohttp import ClientResponseError
+from tessie_api import get_state
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import TessieStatus
+
+# This matches the update interval Tessie performs server side
+TESSIE_SYNC_INTERVAL = 10
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TessieDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching data from the Tessie API."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        api_key: str,
+        vin: str,
+        data: dict[str, Any],
+    ) -> None:
+        """Initialize Tessie Data Update Coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Tessie",
+            update_method=self.async_update_data,
+            update_interval=timedelta(seconds=TESSIE_SYNC_INTERVAL),
+        )
+        self.api_key = api_key
+        self.vin = vin
+        self.session = async_get_clientsession(hass)
+        self.data = self._flattern(data)
+        self.did_first_update = False
+
+    async def async_update_data(self) -> dict[str, Any]:
+        """Update vehicle data using Tessie API."""
+        try:
+            vehicle = await get_state(
+                session=self.session,
+                api_key=self.api_key,
+                vin=self.vin,
+                use_cache=self.did_first_update,
+            )
+        except ClientResponseError as e:
+            if e.status == HTTPStatus.REQUEST_TIMEOUT:
+                # Vehicle is offline, only update state and dont throw error
+                self.data["state"] = TessieStatus.OFFLINE
```

Comment:
This is a calculated state not coming from the API. We don't allow that in integrations that integrate devices or services. Please remove this sensor.

Home Assistant uses the `available` entity property to expose services going offline.

Reply:
> This is a calculated state not coming from the API. We don't allow that in integrations that integrate devices or services. Please remove this sensor.
> 
> Home Assistant uses the `available` entity property to expose services going offline.

Fair enough, I'll move it to a binary sensor since the API does return online/asleep.

Reply:
Thanks!

Reply:
https://github.com/home-assistant/core/pull/105937